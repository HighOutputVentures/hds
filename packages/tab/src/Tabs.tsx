import {
  Box,
  Flex,
  SystemStyleObject,
  Tab,
  TabIndicator,
  TabList,
  TabPanel,
  TabPanels,
  Tabs as ChakraTabs,
} from '@chakra-ui/react';
import * as React from 'react';
import { v4 as uuid } from 'uuid';

type Item = {
  label: string;
  render(): JSX.Element | null;
  badgeCount?: number | string;
};

type Variant = 'solid' | 'plain' | 'underlined';
type Placement = 'start' | 'end' | 'center';
type Orientation = 'vertical' | 'horizontal';

export type TabsProps = {
  items?: Item[];
  variant?: Variant;
  isFitted?: boolean;
  placement?: Placement;
  orientation?: Orientation;
  preferMounted?: boolean;
};

export default function Tabs({
  items,
  variant = 'underlined',
  isFitted,
  placement = 'start',
  orientation = 'horizontal',
  preferMounted,
}: TabsProps) {
  const [selectedIndex, setSelectedIndex] = React.useState(0);

  /*
   * force "horizontal" orientation if variant is "underlined"
   */
  orientation =
    variant === 'underlined' && orientation === 'vertical' ? 'horizontal' : orientation;

  const styles = useStyles({
    variant,
    placement,
    orientation,
  });

  return (
    <ChakraTabs
      orientation={orientation}
      variant="unstyled"
      isFitted={isFitted}
      index={selectedIndex}
      onChange={setSelectedIndex}
      /**
       *
       * User opts to prerender tab contents
       *
       */
      {...(!preferMounted && {
        isLazy: true,
        lazyBehavior: 'unmount',
      })}
      sx={styles.tabs}
      data-testid="hds.tabs"
    >
      <TabList sx={styles.tablist} data-testid="hds.tabs.tablist">
        {items?.map(({ label, badgeCount }, ...args) => (
          <Tab key={uuid()} sx={styles.tab} data-testid="hds.tabs.tab">
            <Box flexGrow={1} textAlign="start">
              {label}
            </Box>

            {badgeCount && (
              <Badge
                count={badgeCount}
                variant={variant}
                isSelected={args[0] === selectedIndex}
              />
            )}
          </Tab>
        ))}
      </TabList>

      {variant === 'underlined' && <TabIndicator sx={styles.tabIndicator} />}

      <TabPanels sx={styles.tabPanels} data-testid="hds.tabs.panels">
        {items?.map(({ render: Component }) => (
          <TabPanel key={uuid()} padding={0} data-testid="hds.tabs.panels.panel">
            <Component />
          </TabPanel>
        ))}
      </TabPanels>
    </ChakraTabs>
  );
}

type UseStylesArg = {
  variant: Variant;
  placement: Placement;
  orientation: Orientation;
};

type UseStylesReturn = {
  tab: SystemStyleObject;
  tabs: SystemStyleObject;
  tablist: SystemStyleObject;
  tabPanels: SystemStyleObject;
  tabIndicator: SystemStyleObject;
};

function useStyles(arg: UseStylesArg) {
  const styles = React.useMemo<UseStylesReturn>(
    () => ({
      tabs: {
        width: 'full',
        margin: '0px',
        padding: '0px',
      },
      tablist: {
        gap: '16px',
        padding: '0px',
        placeContent: placementToPlaceContent(arg.placement),

        ...(arg.orientation === 'vertical' && {
          flexDirection: 'column',
        }),

        ...(arg.orientation === 'horizontal' && {
          flexDirection: 'row',
        }),

        ...(arg.variant === 'underlined' && {
          borderBottom: '1px',
          borderColor: 'neutrals.200',
        }),
      },
      tab: {
        color: 'neutrals.600',
        fontSize: '14px',
        lineHeight: '14px',
        letterSpacing: '0.02em',
        transition: 'colors 300ms ease-in-out',
        fontWeight: 'semibold',
        minWidth: 'auto',
        justifyContent: 'center',
        display: 'flex',
        alignItems: 'center',
        gap: '8px',
        whiteSpace: 'nowrap',

        _hover: {
          color: 'neutrals.700',
        },

        ...(arg.orientation === 'vertical' && {
          minWidth: '145px',
          justifyContent: 'flex-start',
        }),

        ...(arg.variant === 'plain' && {
          height: '32px',
          rounded: '6px',
          paddingX: '12px',
          paddingY: '8px',
          _selected: {
            color: 'neutrals.800',
            bgColor: 'neutrals.100',
          },
        }),

        ...(arg.variant === 'solid' && {
          height: '32px',
          rounded: '6px',
          paddingX: '12px',
          paddingY: '8px',
          _selected: {
            color: 'brand.primary.700',
            bgColor: 'brand.primary.500',
          },
        }),

        ...(arg.variant === 'underlined' && {
          paddingX: '4px',
          paddingTop: '0px',
          paddingBottom: '16px',
          _selected: {
            color: 'brand.primary.700',
          },

          ...(arg.orientation === 'vertical' && {
            paddingBottom: '0px',
          }),
        }),
      },
      tabIndicator: {
        rounded: 'full',
        bgColor: 'brand.primary.700',

        ...(arg.orientation === 'horizontal' && {
          marginTop: '-1px',
          height: '2px',
        }),

        ...(arg.orientation === 'vertical' && {
          width: '2px',
        }),
      },
      tabPanels: {
        width: 'auto',
        height: 'auto',
        padding: '0px',
      },
    }),
    [arg],
  );

  return styles;
}

function placementToPlaceContent(placement?: Placement) {
  switch (placement) {
    case 'center':
      return 'center';
    case 'end':
      return 'flex-end';
    default:
      return 'flex-start';
  }
}

type BadgeProps = {
  count: string | number;
  variant: Variant;
  isSelected?: boolean;
};

function Badge({ count, variant, isSelected }: BadgeProps) {
  return (
    <Flex
      sx={{
        height: 5,
        minWidth: 5,
        paddingX: 1.5,
        color: 'neutrals.800',
        bgColor: 'neutrals.100',
        alignItems: 'center',
        justifyContent: 'center',
        borderRadius: 'lg',
        transition: 'colors 300ms ease-in-out',
        fontSize: '12px',
        lineHeight: '18px',
        fontWeight: 'medium',

        ...(isSelected && {
          ...(variant === 'plain' && {
            color: 'neutrals.900',
            bgColor: 'neutrals.200',
          }),

          ...(variant === 'solid' && {
            color: 'brand.primary.700',
            bgColor: 'blackAlpha.100',
          }),

          ...(variant === 'underlined' && {
            color: 'brand.primary.700',
            bgColor: 'brand.primary.500',
          }),
        }),
      }}
      data-testid="hds.tabs.tab.badge"
    >
      {count}
    </Flex>
  );
}
