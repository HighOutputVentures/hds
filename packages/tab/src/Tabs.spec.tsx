import React from 'react';
import renderer from 'react-test-renderer';
import Tabs from './Tabs';

describe('Snapshot (Tabs)', () => {
  it('Should match snapshot', () => {
    const component = renderer.create(
      <Tabs
        items={[
          {
            label: 'One',
            render() {
              return <div>One</div>;
            },
          },
          {
            label: 'Two',
            render() {
              return <div>Two</div>;
            },
          },
          {
            label: 'Three',
            render() {
              return <div>Three</div>;
            },
          },
        ]}
      />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    component.unmount();
  });
});
