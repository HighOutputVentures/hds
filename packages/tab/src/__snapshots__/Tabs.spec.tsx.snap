// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot (Tabs) Should match snapshot 1`] = `
<div
  className="chakra-tabs css-la4in9"
  data-testid="hds.tabs"
>
  <div
    aria-orientation="horizontal"
    className="chakra-tabs__tablist css-laq73v"
    data-testid="hds.tabs.tablist"
    onKeyDown={[Function]}
    role="tablist"
  >
    <button
      aria-controls="tabs-:r0:--tabpanel--1"
      aria-selected={false}
      className="chakra-tabs__tab css-9btmlu"
      data-testid="hds.tabs.tab"
      id="tabs-:r0:--tab--1"
      onClick={[Function]}
      onFocus={[Function]}
      role="tab"
      tabIndex={-1}
      type="button"
    >
      <div
        className="css-1xedgln"
      >
        One
      </div>
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel--1"
      aria-selected={false}
      className="chakra-tabs__tab css-9btmlu"
      data-testid="hds.tabs.tab"
      id="tabs-:r0:--tab--1"
      onClick={[Function]}
      onFocus={[Function]}
      role="tab"
      tabIndex={-1}
      type="button"
    >
      <div
        className="css-1xedgln"
      >
        Two
      </div>
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel--1"
      aria-selected={false}
      className="chakra-tabs__tab css-9btmlu"
      data-testid="hds.tabs.tab"
      id="tabs-:r0:--tab--1"
      onClick={[Function]}
      onFocus={[Function]}
      role="tab"
      tabIndex={-1}
      type="button"
    >
      <div
        className="css-1xedgln"
      >
        Three
      </div>
    </button>
  </div>
  <div
    className="chakra-tabs__tab-indicator css-131u3bt"
    style={
      {
        "left": 0,
        "position": "absolute",
        "transitionDuration": "0ms",
        "transitionProperty": "left, right, top, bottom, height, width",
        "transitionTimingFunction": "cubic-bezier(0, 0, 0.2, 1)",
        "width": 0,
      }
    }
  />
  <div
    className="chakra-tabs__tab-panels css-pwpux2"
    data-testid="hds.tabs.panels"
  >
    <div
      aria-labelledby="tabs-:r0:--tab-0"
      className="chakra-tabs__tab-panel css-47dblg"
      data-testid="hds.tabs.panels.panel"
      hidden={false}
      id="tabs-:r0:--tabpanel-0"
      role="tabpanel"
      tabIndex={0}
    >
      <div>
        One
      </div>
    </div>
    <div
      aria-labelledby="tabs-:r0:--tab-1"
      className="chakra-tabs__tab-panel css-47dblg"
      data-testid="hds.tabs.panels.panel"
      hidden={true}
      id="tabs-:r0:--tabpanel-1"
      role="tabpanel"
      tabIndex={0}
    />
    <div
      aria-labelledby="tabs-:r0:--tab-2"
      className="chakra-tabs__tab-panel css-47dblg"
      data-testid="hds.tabs.panels.panel"
      hidden={true}
      id="tabs-:r0:--tabpanel-2"
      role="tabpanel"
      tabIndex={0}
    />
  </div>
</div>
`;
